### PR TITLE
[ENH] Nipype-fy ``antsBrainExtraction.sh``

### PR DESCRIPTION
**Pull request by oesteban**
This will enable several steps towards generalization:

  - Play with the template used as target for brain extraction (e.g. macaque, mice)
  - Have more control over settings (important for templates of other species)
  - Use the transforms to the target template (e.g. to initialize the spatial normalization)
  - Revise the brain mask refinements
  - Closely control over what ANTs is doing to headers.
  - (probably we will not notice) leverage nipype parallelization and caching
  - Refine ATROPOS results and merge results with FAST (smoother segmentations)

EDIT: **PLEASE NOTE** that this PR will change how the templates are accessed, trying to normalize naming. So far only two templates are supported, and their tree is like:

```Tree
tpl-MNI152NLin2009cAsym
├── CHANGES
├── template_description.json
├── tpl-MNI152NLin2009cAsym_space-MNI_res-01_brainmask.nii.gz
├── tpl-MNI152NLin2009cAsym_space-MNI_res-01_class-CSF_probtissue.nii.gz
├── tpl-MNI152NLin2009cAsym_space-MNI_res-01_class-GM_probtissue.nii.gz
├── tpl-MNI152NLin2009cAsym_space-MNI_res-01_class-WM_probtissue.nii.gz
├── tpl-MNI152NLin2009cAsym_space-MNI_res-01_label-carpet_atlas.nii.gz
├── tpl-MNI152NLin2009cAsym_space-MNI_res-01_label-eyemask_roi.nii.gz
├── tpl-MNI152NLin2009cAsym_space-MNI_res-01_label-facemask_roi.nii.gz
├── tpl-MNI152NLin2009cAsym_space-MNI_res-01_label-headmask_roi.nii.gz
├── tpl-MNI152NLin2009cAsym_space-MNI_res-01_PD.nii.gz
├── tpl-MNI152NLin2009cAsym_space-MNI_res-01_T1w_brain.nii.gz
├── tpl-MNI152NLin2009cAsym_space-MNI_res-01_T1w.nii.gz
├── tpl-MNI152NLin2009cAsym_space-MNI_res-01_T2w.nii.gz
├── tpl-MNI152NLin2009cAsym_space-MNI_res-02_brainmask.nii.gz
├── tpl-MNI152NLin2009cAsym_space-MNI_res-02_class-CSF_probtissue.nii.gz
├── tpl-MNI152NLin2009cAsym_space-MNI_res-02_class-GM_probtissue.nii.gz
├── tpl-MNI152NLin2009cAsym_space-MNI_res-02_class-WM_probtissue.nii.gz
├── tpl-MNI152NLin2009cAsym_space-MNI_res-02_label-carpet_atlas.nii.gz
├── tpl-MNI152NLin2009cAsym_space-MNI_res-02_label-eyemask_roi.nii.gz
├── tpl-MNI152NLin2009cAsym_space-MNI_res-02_label-facemask_roi.nii.gz
├── tpl-MNI152NLin2009cAsym_space-MNI_res-02_label-headmask_roi.nii.gz
├── tpl-MNI152NLin2009cAsym_space-MNI_res-02_PD.nii.gz
├── tpl-MNI152NLin2009cAsym_space-MNI_res-02_T1w_brain.nii.gz
├── tpl-MNI152NLin2009cAsym_space-MNI_res-02_T1w.nii.gz
└── tpl-MNI152NLin2009cAsym_space-MNI_res-02_T2w.nii.gz
tpl-OASIS30ANTs
├── CHANGES
├── template_description.json
├── tpl-OASIS30ANTs_res-01_brainmask.nii.gz
├── tpl-OASIS30ANTs_res-01_class-brainmask_probtissue.nii.gz
├── tpl-OASIS30ANTs_res-01_class-bstem_probtissue.nii.gz
├── tpl-OASIS30ANTs_res-01_class-Cerebellum_probtissue.nii.gz
├── tpl-OASIS30ANTs_res-01_class-CGM_probtissue.nii.gz
├── tpl-OASIS30ANTs_res-01_class-CSF_probtissue.nii.gz
├── tpl-OASIS30ANTs_res-01_class-DGM_probtissue.nii.gz
├── tpl-OASIS30ANTs_res-01_class-WM_probtissue.nii.gz
├── tpl-OASIS30ANTs_res-01_label-BrainCerebellumExtraction_roi.nii.gz
├── tpl-OASIS30ANTs_res-01_label-BrainCerebellumRegistration_roi.nii.gz
├── tpl-OASIS30ANTs_res-01_T1w_brain.nii.gz
├── tpl-OASIS30ANTs_res-01_T1w.nii.gz
├── tpl-OASIS30ANTs_res-01_variant-4_dtissue.nii.gz
└── tpl-OASIS30ANTs_res-01_variant-6_dtissue.nii.gz
```

I've also taken time to clean all these files up, fixing appropriate dtypes, filling wrong headers, copying the sforms to the qforms, etc.

### CODE REVIEW

File: niworkflows/interfaces/ants.py [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+    output_transform = traits.File(
+        'initialization.mat', usedefault=True, argstr='-o %s',
+        desc='output file name')
+
+
+class AIOuputSpec(base.TraitedSpec):
+    output_transform = traits.File(exists=True, desc='output file name')
+
+
+class AI(ANTSCommand):
```
Comment (effigies) [inline niworkflows/interfaces/ants.py]: Is this an initial affine registration? Didn't we already have something like `AffineInitializer`?

In any event, it should be clearer what `antsAI` does.

File: niworkflows/interfaces/ants.py [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+                                 desc='outside value')
+
+
+class ThresholdImageOutputSpec(base.TraitedSpec):
+    output_image = traits.File(exists=True, desc='resampled file')
+
+
+class ThresholdImage(ANTSCommand):
+    """
+    Resamples an image with a given spacing
```
Comment (effigies) [inline niworkflows/interfaces/ants.py]: This seems mis-applied.

File: niworkflows/data/utils.py [hunk 3]
```diff
@@ -237,7 +237,7 @@ def _get_dataset_dir(dataset_name, data_dir=None, default_paths=None,
     # Search possible system paths
     if default_paths is not None:
         for default_path in default_paths:
-            paths.extend([(d, True) for d in default_path.split(os.pathsep)])
+            paths.extend([(d, True) for d in default_paths.split(os.pathsep)])
```
Comment (effigies) [inline niworkflows/data/utils.py]: I think this is probably a mistake.

File: niworkflows/anat/ants.py [hunk 4]
```diff
@@ -0,0 +1,555 @@
...
+def _select_labels(in_segm, labels):
+    import numpy as np
+    import nibabel as nb
+    from niworkflows.nipype.utils.filemanip import fname_presuffix
+
+    out_files = []
+
+    nii = nb.load(in_segm)
+    for l in labels:
+        data = np.zeros(nii.shape)
```
Comment (effigies) [inline niworkflows/anat/ants.py]: `dtype=np.uint8`

File: niworkflows/anat/ants.py [hunk 5]
```diff
@@ -0,0 +1,555 @@
...
+)
+
+ATROPOS_MODELS = {
+    'T1': (3, 1, 2, 3),
+    'T2': (3, 3, 2, 1),
+    'FLAIR': (3, 1, 3, 2),
+}
+
+
+def brain_extraction(name='antsBrainExtraction',
```
Comment (effigies) [inline niworkflows/anat/ants.py]: Should we stick with our convention `init_brain_extraction_wf`?

File: niworkflows/anat/ants.py [hunk 6]
```diff
@@ -0,0 +1,555 @@
...
+        (dil_brainmask, get_brainmask, [('output_image', 'op1')]),
+        (inu_n4, apply_mask, [('output_image', 'in_file')]),
+        (get_brainmask, apply_mask, [('output_image', 'mask_file')]),
+        (get_brainmask, outputnode, [('output_image', 'out_mask')]),
+        (apply_mask, outputnode, [('out_file', 'bias_corrected')]),
+        (inu_n4, outputnode, [('bias_image', 'bias_image')]),
+    ])
+
+    if atropos_refine:
+        # TODO: connect output segmentation to 'outputnode.out_segm'
```
Comment (effigies) [inline niworkflows/anat/ants.py]: Looks like this has been done?

File: niworkflows/interfaces/ants.py (hunk 1 above)
Comment (oesteban) [reply niworkflows/interfaces/ants.py]: It is a replacement for AffineInitializer as of ANTs 2.1.0.

File: niworkflows/anat/ants.py (hunk 5 above)
Comment (oesteban) [reply niworkflows/anat/ants.py]: sounds good

File: niworkflows/anat/ants.py (hunk 6 above)
Comment (oesteban) [reply niworkflows/anat/ants.py]: Ups, yeah

File: niworkflows/data/utils.py (hunk 3 above)
Comment (oesteban) [reply niworkflows/data/utils.py]: Yep

**Comment by effigies** [comment]
Getting to this now. Sorry about the delay.

**Comment by oesteban** [comment]
Actually, besides @effigies comments, it would be wonderful to have an overall opinion from @mgxd as our expert on templates and @satra as this workflow is a good candidate for the inauguration of https://github.com/niflows. WDYT?

File: niworkflows/anat/ants.py [hunk 7]
```diff
@@ -0,0 +1,559 @@
...
+    mrg_tmpl.inputs.in1 = tpl_target_path
+    mrg_target = pe.Node(niu.Merge(2), name='mrg_target')
+
+    # Initialize transforms with antsAI
+    init_aff = pe.Node(AI(
+        metric=('Mattes', 32, 'Regular', 0.2),
+        transform=('Affine', 0.1),
+        search_factor=(20, 0.12),
+        principal_axes=False,
+        convergence=(10, 1e-6, 10),
```
[the remark this answers is not in the document]
Comment (oesteban) [reply niworkflows/anat/ants.py]: Yes. I was lazy and did not want to code a convergence trait with a parser that can use the single value config (`-c 10`). So `1e-6` and `10` are the defaults of `antsAI` in the single value setting. When we upstream the interfaces to nipype, we should take care of that.

Thus, `-c 10` is equivalent to `-c [10, 1e-6, 10]`.

> I haven't found the correlate of `apply_mask` in the original

One of the reasons why I wanted to translate this workflow is readability, precisely. The equivalent to `apply_mask` is: https://github.com/ANTsX/ANTs/blob/59d6bd2a23f55a08ca2f88224cfed3c1aac7f393/Scripts/antsBrainExtraction.sh#L577

**Comment by satra** [comment]
@effigies  and @oesteban - i was hoping to write a reprozip tool and use the graph to help create nipype workflows. perhaps we can chat about this at ohbm. (btw, i posted two nipype projects there).

that being said, ants is going to be particularly sensitive in mapping command line expressions to nipype interface flags, just because of the way ants command lines are often structured. so this may end up being much more work. however, we should be able to automatically construct much of the graph, not the details, which we may be able to adjust by hand, or write an adapter.

i also have a workflow for joint fusion labeling on the basis of antscorticalthickness.sh output, which i think internally calls this script. here is the gist: https://gist.github.com/satra/ca140fe2cb41659b1b9666909df19cf4 and that should also be a candidate for a niflow!

**Comment by oesteban** [comment]
Thumbs up?

**Comment by effigies** [comment]
Sorry, still need to check the Atropos workflow.

File: niworkflows/anat/ants.py [hunk 8]
```diff
@@ -0,0 +1,559 @@
...
+from ..interfaces.fixes import (
+    FixHeaderRegistration as Registration,
+    FixHeaderApplyTransforms as ApplyTransforms,
+)
+
+ATROPOS_MODELS = {
+    'T1': (3, 1, 2, 3),
+    'T2': (3, 3, 2, 1),
+    'FLAIR': (3, 1, 3, 2),
+}
```
Comment (effigies) [inline niworkflows/anat/ants.py]: This is opaque. At the least I would specify that this is `(K, CSF, GM, WM)`, but we could also do something a bit more structured, like:

```Python
ATROPOS_MODELS = {
   'T1': {'wm': 3, 'gm': 2, 'csf': 1},
   'T2': {'wm': 1, 'gm': 2, 'csf': 3},
   'FLAIR': {'wm': 2, 'gm': 3, 'csf': 1},
}
```

Or:

```Python
ATROPOS_MODELS = {
   'T1': ['csf', 'gm', 'wm'],
   'T2': ['wm', 'gm', 'csf'],
   'FLAIR': ['csf', 'wm', 'gm'],
}
```

And then we can end up with something a little more clear in the `_select_labels`, as well. Just a thought, and maybe one for another day.

File: niworkflows/anat/ants.py [hunk 9]
```diff
@@ -0,0 +1,559 @@
...
+        (md_7, fill_7, [('output_image', 'op1')]),
+        (fill_7, add_7_2, [('output_image', 'op1')]),
+        (pad_mask, add_7_2, [('output_image', 'op2')]),
+        (add_7_2, md_7_2, [('output_image', 'op1')]),
+        (md_7_2, me_7_2, [('output_image', 'op1')]),
+        (me_7_2, depad_mask, [('output_image', 'op1')]),
+        (add_gm_wm, depad_segm, [('output_image', 'op1')]),
+        (relabel_wm, depad_wm, [('output_product_image', 'op1')]),
+        (relabel_gm, depad_gm, [('output_product_image', 'op1')]),
+        (sel_labels, depad_csf, [('out_csf', 'op1')]),
```
Comment (effigies) [inline niworkflows/anat/ants.py]: Outputs of `depad_segm`, `depad_wm`, `depad_gm`, `depad_csf` are unused. Are these supposed to be in `outputnode.out_segm` and `outputnode.out_tpms`?

File: niworkflows/anat/ants.py [hunk 10]
```diff
@@ -0,0 +1,559 @@
...
+        (depad_mask, outputnode, [('output_image', 'out_mask')]),
+    ])
+    return wf
+
+
+def _list(in_files):
+    if isinstance(in_files, (bytes, str)):
+        return [in_files]
+    return in_files
```
[the remark this answers is not in the document]
Comment (oesteban) [reply niworkflows/anat/ants.py]: Turns out that `_list` was not even used.

File: niworkflows/anat/ants.py (hunk 8 above)
Comment (oesteban) [reply niworkflows/anat/ants.py]: I've switched to OrderedDict for the specification, but I'm +1 in using plain dicts and modifying `_select_labels`

File: niworkflows/anat/ants.py (hunk 9 above)
Comment (oesteban) [reply niworkflows/anat/ants.py]: They were missing, you are right.

**Pull request merged**